nav2_recoveries/plugins/assisted_teleop.cpp

```diff
@@ -0,0 +1,300 @@
+{
+  speed_x = msg->linear.x;
+  speed_y = msg->linear.y;
+  angular_vel_ = msg->angular.z;
+  if (go) {
+    if (!checkCollision()) {
+      RCLCPP_INFO(logger_, "Reducing velocity by a factor of %.2f", scaling_factor);
+    }
+    moveRobot();
+  }
+}
+
+bool
+AssistedTeleop::updatePose()
+{
+  if (!nav2_util::getCurrentPose(
+      current_pose, *tf_, global_frame_, robot_base_frame_,
+      transform_tolerance_))
+  {
+    RCLCPP_ERROR(logger_, "Current robot pose is not available.");
+    return false;
+  }
+  projected_pose.x = current_pose.pose.position.x;
+  projected_pose.y = current_pose.pose.position.y;
+  projected_pose.theta = tf2::getYaw(
+    current_pose.pose.orientation);
+  return true;
+}
+
+void
+AssistedTeleop::projectPose(
+  double speed_x, double speed_y,
+  double angular_vel_, double projection_time)
+{
+  // Project Pose by time increment
+  projected_pose.x += projection_time * (
+    speed_x * cos(projected_pose.theta));
+  projected_pose.y += projection_time * (
+    speed_y * sin(projected_pose.theta));
+  projected_pose.theta += projection_time *
+    angular_vel_;
+}
+
+bool
+AssistedTeleop::checkCollision()
+{
+  double hypot = std::hypot(speed_x, speed_y, angular_vel_);
+  const double dt =
+    (hypot != 0) ? (costmap_ros_->getResolution() / std::fabs(hypot)) : projection_time_;
+  int loopcount = 1;
+
+  while (true) {
+    if (updatePose()) {
+      double time_to_collision = loopcount * dt;
+      if (time_to_collision >= projection_time_) {
+        scaling_factor = 1;
+        break;
+      }
+      scaling_factor = projection_time_ / (time_to_collision);
+      loopcount++;
+
+      projectPose(speed_x, speed_y, angular_vel_, time_to_collision);
+
+      if (!collision_checker_->isCollisionFree(projected_pose)) {
+        RCLCPP_WARN(logger_, "Collision approaching in %.2f seconds", time_to_collision);
+        return false;
+      }
+    }
+  }
+  return true;
+}
+
+void
+AssistedTeleop::moveRobot()
+{
+  auto cmd_vel = std::make_unique<geometry_msgs::msg::Twist>();
+
+  cmd_vel->linear.x = speed_x / scaling_factor;
+  cmd_vel->linear.y = speed_y / scaling_factor;
+  cmd_vel->angular.z = angular_vel_ / scaling_factor;
+  double hypot = std::hypot(cmd_vel->linear.x, cmd_vel->linear.y, cmd_vel->angular.z);
+
+  if (std::fabs(hypot) < linear_velocity_threshold_) {
+    stopRobot();
+  } else {
+    vel_pub_->publish(std::move(cmd_vel));
+  }
+}
+
+void AssistedTeleop::execute()
+{
+  RCLCPP_INFO(logger_, "Attempting %s", recovery_name_.c_str());
+
+  if (!enabled_) {
+    RCLCPP_WARN(
+      logger_,
+      "Called while inactive, ignoring request.");
+    return;
+  }
+
+  // Log a message every second
+  {
+    auto node = node_.lock();
+    if (!node) {
+      throw std::runtime_error{"Failed to lock node"};
+    }
+
+    auto timer = node->create_wall_timer(
+      1s,
+      [&]()
+      {RCLCPP_INFO(logger_, "%s running...", recovery_name_.c_str());});
+  }
+
+  auto start_time = steady_clock_.now();
```

Comment:
You locked the node in this function, why not just do `node->now()` and remove the clock variable?